R/is_reproducible.R

```diff
@@ -0,0 +1,40 @@
+#' Reproducibility Check for `@code` Slot in `teal_data`
+#'
+#' Validates the reproducibility of the `teal_data@env` environment using the `@code` slot.
+#'
+#' @return If the objects returned by the `@code` slot in the input `teal_data` object exactly match
+#' the objects in `teal_data@env`, the function updates the `@valid` slot to `TRUE` in the `teal_data` object.
+#' If the `@code` fails to recreate objects in `teal_data@env`, an error is thrown.
+#'
+#' @param teal_data `teal_data` object
+#' @examples
+#' tdata1 <- teal_data()
+#' tdata1 <- within(tdata1, {
+#'   a <- 1
+#'   b <- a^5
+#'   c <- list(x = 2)
+#' })
+#' is_reproducible(tdata1)
+#'
+#' tdata2 <- teal_data(x1 = iris, code = "x1 = iris")
+#' is_reproducible(tdata2)
+#' is_reproducible(tdata2)@valid
+#' tdata2@valid
+#'
+#' @export
+is_reproducible <- function(teal_data) {
+  checkmate::assert_class(teal_data, "teal_data")
+  if (teal_data@valid) {
+    return(teal_data)
+  }
+  eval_env <- new.env()
+  eval(parse(text = teal_data@code), envir = eval_env)
+
+  reproducible <- all.equal(teal_data@env, eval_env)
+  if (reproducible) {
+    teal_data@valid <- TRUE
+    teal_data
+  } else {
+    stop("Code validation failed.")
+  }
```

Comment:
I was thinking if this simpler or more robust.
```suggestion
  new_teal_data <- teal_data() |> eval_code(teal_data@code)
  datanames(new_teal_data) <- datanames(teal_data)
  join_keys(new_teal_data) <- join_keys(new_teal_data)
  
  reproducible <- all.equal(teal_data@env, new_teal_data@env)
  if (isTRUE(reproducible)) {
    new_teal_data
  } else {
    stop("Code validation failed.")
  }
```

Reply:
Thanks @gogonzo I see that with `teal_data()` you are able to set `@valid = TRUE` just because `code` and `datanames` are empty :) Also very clever to use our own `eval_code` instead of customized `eval(parse(text=`.

I would simplify this a bit, and would just modify `@valid` field for the old object, and would pass `datanames` and `join_keys` to the new one. This reduced 2 lines into 1. What about below

```r
new_teal_data <- teal_data() |> eval_code(teal_data@code)

reproducible <- all.equal(teal_data@env, new_teal_data@env)
if (isTRUE(reproducible)) {
  teal_data@valid <- TRUE
  teal_data
} else {
  stop("Code validation failed.")
}
```

Reply:
Even better.

Reply:
https://github.com/insightsengineering/teal.data/pull/188/commits/873283de17bcc63092720713316eacab75da987c